Switch to Ruby 3.0.0

citizen428 opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

## What type of PR is this? (check all applicable)

- [X] Update

## Description

This is a tracking PR to see what we have to do to get the app running with Ruby 3.

### Issues

* [ ] [`fix-db-schema-conflicts` ](https://github.com/jakeonrails/fix-db-schema-conflicts) specifies a requirde Ruby version of `'~> 2.0'` in its `Gemfile`. The gem hasn't been updated in almost 2 years, so we may have to fork it but can try to submit a PR first. It's a dev dependency only.
* [ ] The[ `katex` gem](https://github.com/glebm/katex-ruby) (used for the katex Liquid tag) has `'~> 2.3'` as required Ruby version. It hasn't been updated since August 2019, so the same as above applies.
* [ ] We need a container build for Ruby 3.0.0 (cc @jdoss)

## Related Tickets & Documents

n/a

## QA Instructions, Screenshots, Recordings

All tests should pass and the app should work.

### UI accessibility concerns?

n/a

## Added tests?

- [X] No, and this is why: this is not adding new functionality, but upgrading Ruby, so we'd expect existing tests to still pass.

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [X] README
- [] No documentation needed

jdoss commented:
I'm tracking Ruby 3 in Fedora and I expect that it won't make it into F33. I'll probably make a `forem/ruby:3.0.0` container so we don't have to wait for Fedora 34 to come out approximately around May. I'll see if I can get this done this week.

citizen428 commented:
@jdoss Seems like the container build on Buildkite failed with a  Nokogiri error. Will you look into that or should I?

rhymes commented:
+1 on forking `fix-db-schema-conflicts` to `forem/` and +1 to removing `binding_of_caller`

citizen428 commented:
@jdoss This seems to be unrelated, `cld` does not depend on `nokogiri`. Can you please run the suggested command in the container and paste the output here? If you don't have time for that I'll look into it myself, but it may be a bit before I get around to that.

```
gem install cld -v '0.8.0' --source 'https://rubygems.org/'
```

citizen428 commented:
I don't have permission to fork repositories into the `forem` org, so for now I forked `fix_db_schema_conflicts` into my account and will transfer it later.

jdoss commented:
> @jdoss This seems to be unrelated, `cld` does not depend on `nokogiri`. Can you please run the suggested command in the container and paste the output here? If you don't have time for that I'll look into it myself, but it may be a bit before I get around to that.
> 
> ```
> gem install cld -v '0.8.0' --source 'https://rubygems.org/'
> ```

Yea it is unrelated. I did `gem install cld -v '0.8.0' --source 'https://rubygems.org/'` yesterday and it resulted in the same error that I posted above. It seems to be a specific issue with this gem and Ruby 3.

joshpuetz commented:
The `cld` gem is old as dirt and not maintained anymore: I wouldn't be surprised if it requires forking and a gemspec change to enable it to install on Ruby 3. (Background: it's used for detecting language).

This gem is also a blocker for running Forem under Arm64 on Apple Silicon. @fdoxyz experimented with replaced it with the `cld2-small` gem (https://rubygems.org/gems/cld2-small) when he got Forem running on Raspberry Pi (https://forem.dev/fdoxyz/how-to-get-forem-to-run-on-arm-5hea). Could be worth switching to it.

citizen428 commented:
@joshpuetz I had checked, it's not gemspec related (you wouldn't see a compilation error in that case, but a bundler error). Thanks for the hint about `cld2-small`, that sounds promising.

citizen428 commented:
I switched this branch to use `cld2-small`.

citizen428 commented:
Before I dig into Travis myself, does anyone know why all runs there fail completely with what looks like Knapsack errors?

rhymes commented:
Looking at https://travis-ci.com/github/forem/forem/jobs/471977202#L2545 here it looks like  https://github.com/forem/forem/blob/c2c1366a4ddc452f9d7e587d1b930391acbe0c06/spec/support/initializers/ci_csv_formatter.rb#L67 is not valid anymore. This works: `CSV.open(csv_filename, "w", **with_headers)`

rhymes commented:
There's a segfault on `sassc` in the buildkite build: 

> /opt/apps/forem/vendor/bundle/ruby/3.0.0/gems/sassc-2.4.0/lib/sassc/engine.rb:43: [BUG] Segmentation fault at 0x00007f6900000000

I also added an empty commit to trigger Travis which seems stuck

citizen428 commented:
@rhymes Yeah, I noticed this too, but had no time to investigate yet 😢

joshpuetz commented:
@citizen428 @rhymes I found this similar report or segfaults in SASCC that looks to be in the same place: https://github.com/sass/sassc-ruby/issues/197

The favored workaround seems to be:

```
Rails.application.config.assets.configure do |env|
  env.export_concurrent = false
end
```

citizen428 commented:
@joshpuetz Thanks for this, unfortunately it still segfaults. 😢

citizen428 commented:
@jdoss Another problem I notice in the container build is that `dnf` now installs node 14.15 by default, but we are on 14.14 LTS.

```sh
error dev.to@1.0.0: The engine "node" is incompatible with this module. Expected version "14.14.x". Got "14.15.4"
error Found incompatible module
```

I started fiddling around with that, but will probably continue some other day.

rhymes commented:
According to https://nodejs.org/en/ - the LTS is 14.15.x now.

I wonder if there's a way for us to get notified about these changes

jdoss commented:
@citizen428 I wouldn't consider it a problem but a feature. IMO it is a better to stay lock step with whatever `node` version Fedora ships. It shouldn't change much outside of major Fedora version changes and when it does it is a forcing function for us to upgrade too. 

The Ruby 3.0 yakshave continues!

![](https://xo.tc/images/yackshaving-lightbulb.gif)

citizen428 commented:
@jdoss I have no problem staying in lockstep with Fedora, I just find it a bit weird how short the "active" phase of Node LTS versions seems to be.

citizen428 commented:
@jdoss I have no problem in staying in lockstep with Fedora, I'm questioning if we're doing things correctly. If you look at the LTS graphic I posted above, they are for major versions (v14, v15, etc), but we lock to minor versions. Maybe we can relax our `.nvmrc` a bit here, cc @nickytonline.

nickytonline commented:
Yeah I'm fine with relaxing it.

citizen428 commented:
Got CodeClimate working again by updating the Rubocop version on their end:

```yaml
plugins:
  rubocop:
    enabled: true
    channel: rubocop-1-10
```

citizen428 commented:
The next problem is that the `Slack::Notifier` gem is no longer maintained and currently doesn't work with Ruby 3 without monkey-patches.

rhymes commented:
> The next problem is that the Slack::Notifier gem is no longer maintained and currently doesn't work with Ruby 3 without monkey-patches.

time to find an alternative :D

citizen428 commented:
@rhymes Agreed, but it may be a bit before I get to work on this branch again, my OSS week is over.

maestromac commented:
I'm not sure what it is, but after merging master, I can not run any `bin/rails DB:__` action. Travis [builds](https://travis-ci.com/github/forem/forem/builds/221073613) are running into this too. Ruby 3 would hang and I won't be able to exit out of it. Git bisect said it's 7c31a9cc9882cd7bc58f0a69f5817f0986e835d7 that caused this but I have yet to figure out what part of it causes `bin/setup` to hang.

citizen428 commented:
@maestromac Interesting, since this doesn't happen on `master` it really seems to be related to the Ruby 3 upgrade here.

The first things I tried were related to the usual suspects, e.g. `spring` but that got me nowhere. There also wasn't anything interesting in the logs, except for an unrelated (because this also happens on `master`) deprecation warning that we should fix sooner than later:

> DEPRECATION WARNING: Initialization autoloaded the constants RateLimitChecker, ForemStatsDriver, ForemStatsDrivers, ActsAsForemStatsDriver, and ForemStatsDrivers::DatadogDriver.
> 
> Being able to do this is deprecated. Autoloading during initialization is going
> to be an error condition in future versions of Rails.
> 
> Reloading does not reboot the application, and therefore code executed during
> initialization does not run again. So, if you reload RateLimitChecker, for example,
> the expected changes won't be reflected in that stale Class object.
> 
> These autoloaded constants have been unloaded.
> 
> Please, check the "Autoloading and Reloading Constants" guide for solutions.

The tl;dr of this is that we're triggering autoloading from initializers, which has the potential for loaded classes to go stale, see https://github.com/rails/rails/issues/36363.

Back to the original issue:

I then tried to run rails with the `--trace` (`-t`) option, but that wasn't super helpful:

```ruby
$ rails -t db:drop
** Invoke db:drop (first_time)
** Invoke db:load_config (first_time)
** Invoke environment (first_time)
** Execute environment
```

It just hangs, pegging one CPU core and you basically have to kill it with fire (`SIGKILL`, not `SIGTERM`). If you do it early enough `Ctrl-C` still works, but them we get stuck in a loop.

I then started googling, which led me to [this SO thread](https://stackoverflow.com/questions/65829106/app-rails-6-1-1-cant-boot-after-upgrading-to-ruby-3-0) and a similar one I can't find anymore. No solutions, but people experiencing similar issues (app hanging, can't `Ctrl-C` out of it). 

So I brought out the heavier guns in the form of `strace`, but didn't see anything immediately obvious (it seemed at one point to hang while trying to allocate more memory, but I'm quite sure that was a red herring). I then went and attached `gdb` to the running process so I could trigger a backtrace from another terminal:

```
$ gdb -p <pid>

(gdb) call (void) rb_backtrace()
```

The backtrace is very long, I dumped it in [a gist](https://gist.github.com/citizen428/1133d91176a4dad8fc690e03e182f451).

I have some errands to run now, I can look at this again later.

citizen428 commented:
I put a breakpoint in the last initializer and then started stepping from there, which eventually ended up hanging in this block:

```ruby
From: /home/michi/.asdf/installs/ruby/3.0.0/lib/ruby/3.0.0/tsort.rb:349 TSort.each_strongly_connected_component:

    342: def TSort.each_strongly_connected_component(each_node, each_child) # :yields: nodes
    343:   return to_enum(__method__, each_node, each_child) unless block_given?
    344:
    345:   id_map = {}
    346:   stack = []
    347:   each_node.call {|node|
    348:     unless id_map.include? node
 => 349:       TSort.each_strongly_connected_component_from(node, each_child, id_map, stack) {|c|
    350:         yield c
    351:       }
    352:     end
    353:   }
    354:   nil
    355: end
```

IIRC the topolical sort here is used for sorting the dependency graph. I tried a couple of things including upgrading Rubygems itself, but that didn't help either.

I'm probably overthinking it at this point, so I'm gonna step away from it for a bit.

maestromac commented:
Thanks for chiming in @citizen428 ! 

I tried digging further based on the information you gave me but I got nowhere. Putting `binding.pry` directly in `ruby/3.0.0/tsort.rb:349` also got me nowhere. btw how were you able to debug/locate this tsort.rb? I tried with `binding.pry` directly in `bin/setup` but It just gets stuck immediately.

maestromac commented:
Re-adding Buffer gem fixed the issue... :thinking: :man_shrugging: . Additionally,

- Adding Buffer only to the test environment does not solve the problem.
- Adding only Buffer gem's dependency (environ, faraday_middleware, and yajl-ruby) does not solve the problem.
- Downgrading Bundler doesn't solve the problem either.

Did we stumble upon a Ruby 3 bug?

maestromac commented:
I am now blocked. I need to wait for Rails 6.1 before I can continue. Rail's caching with Redis is busted.

citizen428 commented:
>  btw how were you able to debug/locate this tsort.rb? I tried with binding.pry directly in bin/setup but It just gets stuck immediately.

`binding.pry` at the end of the last initializer (Zeitwerk) and then stepping through the code until I got stuck there.

> Re-adding Buffer gem fixed the issue... 

This is bizarre, we have no problem on `master` with this. 

> Rail's caching with Redis is busted.

Can you elaborate a bit? Is this related to Ruby 3?

maestromac commented:
@citizen428 yeah, it's keyword argument issue. Try `Rails.cache.write("test", 1) in console and you'll get incorrect number of arguments. This isn't a problem when I tried it on @rhymes 's Rails 6.1 branch with Ruby 3.

citizen428 commented:
Ugh, this problem really nerd-sniped me. More things I tried that didn't help (after removing the buffer gem again):

1. Removing the Bootsnap cache
2. Disabling Bootsnap completely in `config/boot.rb`
3. Rubygems and Bundler updates/maintenance: `gem update --system`, `bundle update --bundler`, `gem pristine --all`
4. `rbspy` (doesn't work with Ruby 3 yet)

Re-adding the gem makes things work again 🤯 

I then created a new branch off `master`, changed `.ruby-version`, and removed `binding_of_caller`: same behavior, it just hung. Adding the buffer gem to this new branch fixed things too. 🤷

jgaskins commented:
@maestromac Oooh, is it because Ruby 3 removes the hash-to-kwargs functionality that was deprecated in Ruby 2.7? https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/

maestromac commented:
@jgaskins most of the spec failure here is indeed that, it's just that any caching-redis hash-to-kwargs issue is resolved via Rails update. The other issue is our rails app completely hangs if we remove buffer gem from our `Gemfile` which is completely bizarre.

maestromac commented:
Returning to this weird issue again.

I started a fresh rails 6.1.3 app with Ruby 3. Copied the whole Gemfile over and was able to replicate the issue.
- If I remove buffer gem, Ruby hangs.
- If I remove buffer gem AND imgproxy gem, the issue is resolved

I got this result from commenting out Gemfile half at a time slowing combing through it. The result is mixed because I can keep imgproxy gem but I would need to rmove omniauth instead. It's all very very confusing.

It's potentially this issue https://github.com/rails/rails/issues/38666 but I have couldn't verify it

citizen428 commented:
@maestromac I at least don't have `:bin:` in my path, so that's probably not it. That said, I had recently pulled down this branch again, updated all gems and was able to boot the app, but every request ended with "Bad Request" so I quickly abandoned this.

djuber commented:
>  I had recently pulled down this branch again, updated all gems and was able to boot the app

This also worked for me (or "reinstalled buffer and updated all gems" in my case worked).

djuber commented:
I'm just going to leave this here until I can think about it more http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-dev/50999 

I'll point out the removing imgproxy seemed to have unstuck this, that imgproxy depends on anyway_config which refines ::Hash, and that the gdb stack trace is getting stuck in a loop resolving instance methods of Hash or one of its  included modules for me. 

I'm leaning weakly toward something extending Hash or Hash including a module that includes refinements - I initially wanted that to be Hashie (since my gemfile bisection landed me around omniauth-oauth rather than imgproxy) but anyway_config also looks fishy. 

I'll see if I can get a cleaner reproduction for this. 

I'll also see if I can rig up Jeremy Evans suggested patch and confirm that our issue goes away
http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-dev/51032

djuber commented:
applied the patch from http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-dev/51032 (detect 1-cycles in method resolution) and  `bin/setup` worked against the patched ruby (removed rbenv from bashrc and added path to patched ruby first) - re-ran to confirm it failed against the rbenv built ruby 3.0.0 (so this was not due to any fooling with the gemfile).

https://bugs.ruby-lang.org/issues/17494 we may be blocked on this issue if we don't want to dig any deeper into the gems

maestromac commented:
Looks like this hanging issue resolved but not yet released https://bugs.ruby-lang.org/issues/17494#change-92721

joshpuetz commented:
Ruby 3.02 has been released! https://www.ruby-lang.org/en/news/2021/07/07/ruby-3-0-2-released/

citizen428 commented:
@maestromac I took the liberty of updating this branch to 3.0.2. I know we don't have a container image on `quay.io/forem/ruby:3.0.2` yet but I talked to @jdoss and he will put one together once there's an F34 package for Ruby 3.0.2.

I'll also merge back `main` and go through the conflicts.

djuber commented:
just confirming that 3.0.2 contains the backported fix for 17806 (which fixed 17494) 

https://github.com/ruby/ruby/compare/v3_0_1...v3_0_2 
> merge revision(s) 39a2ba5: [Backport #17806]

maestromac commented:
I don't really understand why but the Travis OS build we are using [doesn't have Ruby 3.0.2 yet ](http://rubies.travis-ci.org/). I tried installing it directly and change it to Ubuntu 20.04 but the issue still remains. I'm guessing that 3.0.2 will eventually be available.

citizen428 commented:
@maestromac It's indeed confusing, as it only seems to be missing for x86_64, they have it for other archs, even pretty obscure ones.

maestromac commented:
It's working now. Also, Heroku supports 3.0.2  https://devcenter.heroku.com/changelog-items/2209

maestromac commented:
Build passed :sob:  but BuildKite doesn't have the newer ruby yet?

```
Step 1/20 : FROM quay.io/forem/ruby:3.0.2 as builder
manifest for quay.io/forem/ruby:3.0.2 not found: manifest unknown: manifest unknown
```

jgaskins commented:
@maestromac We own that container repo so we need to build that container. 🙂 https://quay.io/repository/forem/ruby?tag=latest&tab=tags

maestromac commented:
Now that we have access to 3.0.2 build. it's getting hit with new error https://buildkite.com/forem/build-containers/builds/6766#9778a8b3-442d-4c3f-833f-67d0da55e41d/108-1206

>Errno::EPERM: Operation not permitted @ rb_check_realpath_internal - /opt/apps/forem/vendor/bundle/ruby/3.0.0/gems/ruby-next-core-0.12.0/lib/ruby-next/..

maestromac commented:
Please ignore that weird Gitpod integration that suddenly just re-appeared

djuber commented:
took me a minute to see it in the travis logs - but https://app.travis-ci.com/github/forem/forem/jobs/530726032#L329 we did build and test against ruby 3.0.2 as expected

jgaskins reviewed (commented):
I don't seem to be able to successfully run `bundle install` on this branch. It looks like `byebug` isn't able to compile the C extension. It works fine on my Intel Mac, but not on my M1 Mac.

<details>

```
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

current directory:
/Users/jamiegaskins/.rvm/gems/ruby-3.0.2/gems/byebug-11.1.3/ext/byebug
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/bin/ruby -I
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0 -r
./siteconf20210812-93799-9wbjds.rb extconf.rb
creating Makefile

current directory:
/Users/jamiegaskins/.rvm/gems/ruby-3.0.2/gems/byebug-11.1.3/ext/byebug
make DESTDIR\= clean

current directory:
/Users/jamiegaskins/.rvm/gems/ruby-3.0.2/gems/byebug-11.1.3/ext/byebug
make DESTDIR\=
compiling breakpoint.c
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/defines.h:73:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/backward/2/attributes.h:43:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/attr/pure.h:25:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/assert.h:132:1:
error: '__declspec' attributes are not enabled; use '-fdeclspec' or
'-fms-extensions' to enable support for __declspec attributes
RBIMPL_ATTR_NORETURN()
^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/attr/noreturn.h:29:33:
note: expanded from macro 'RBIMPL_ATTR_NORETURN'
# define RBIMPL_ATTR_NORETURN() __declspec(noreturn)
                                ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:24:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/anyargs.h:77:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/intern/vm.h:64:1:
error: '__declspec' attributes are not enabled; use '-fdeclspec' or
'-fms-extensions' to enable support for __declspec attributes
RBIMPL_ATTR_NORETURN()
^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/attr/noreturn.h:29:33:
note: expanded from macro 'RBIMPL_ATTR_NORETURN'
# define RBIMPL_ATTR_NORETURN() __declspec(noreturn)
                                ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:25:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/char.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/int.h:26:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/long.h:70:1:
error: '__declspec' attributes are not enabled; use '-fdeclspec' or
'-fms-extensions' to enable support for __declspec attributes
RBIMPL_ATTR_NORETURN()
^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/attr/noreturn.h:29:33:
note: expanded from macro 'RBIMPL_ATTR_NORETURN'
# define RBIMPL_ATTR_NORETURN() __declspec(noreturn)
                                ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:25:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/char.h:29:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:28:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:46:14:
error: expected parameter declarator
RUBY_ALIGNAS(SIZEOF_VALUE)
             ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/value.h:41:23:
note: expanded from macro 'SIZEOF_VALUE'
# define SIZEOF_VALUE SIZEOF_LONG
                      ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/arm64-darwin20/ruby/config.h:61:21:
note: expanded from macro 'SIZEOF_LONG'
#define SIZEOF_LONG 8
                    ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:25:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/char.h:29:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:28:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:46:14:
error: expected ')'
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/value.h:41:23:
note: expanded from macro 'SIZEOF_VALUE'
# define SIZEOF_VALUE SIZEOF_LONG
                      ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/arm64-darwin20/ruby/config.h:61:21:
note: expanded from macro 'SIZEOF_LONG'
#define SIZEOF_LONG 8
                    ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:46:1:
note: to match this '('
RUBY_ALIGNAS(SIZEOF_VALUE)
^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/backward/2/stdalign.h:27:22:
note: expanded from macro 'RUBY_ALIGNAS'
#define RUBY_ALIGNAS RBIMPL_ALIGNAS
                     ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/stdalign.h:66:44:
note: expanded from macro 'RBIMPL_ALIGNAS'
# define RBIMPL_ALIGNAS(_) __declspec(align(_))
                                           ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:25:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/char.h:29:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:28:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:47:1:
error: expected function body after function declarator
RBasic {
^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:82:23:
error: incomplete definition of type 'struct RBasic'
    return RBASIC(obj)->klass;
           ~~~~~~~~~~~^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:82:12:
note: forward declaration of 'struct RBasic'
    return RBASIC(obj)->klass;
           ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:34:50:
note: expanded from macro 'RBASIC'
#define RBASIC(obj)          RBIMPL_CAST((struct RBasic *)(obj))
                                                 ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:25:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/char.h:29:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:30:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:35:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/value_type.h:157:28:
error: incomplete definition of type 'struct RBasic'
    VALUE ret = RBASIC(obj)->flags & RUBY_T_MASK;
                ~~~~~~~~~~~^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/value_type.h:157:17:
note: forward declaration of 'struct RBasic'
    VALUE ret = RBASIC(obj)->flags & RUBY_T_MASK;
                ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:34:50:
note: expanded from macro 'RBASIC'
#define RBASIC(obj)          RBIMPL_CAST((struct RBasic *)(obj))
                                                 ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:25:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/char.h:29:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:30:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:237:23:
error: incomplete definition of type 'struct RBasic'
    return RBASIC(obj)->flags & flags;
           ~~~~~~~~~~~^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:237:12:
note: forward declaration of 'struct RBasic'
    return RBASIC(obj)->flags & flags;
           ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:34:50:
note: expanded from macro 'RBASIC'
#define RBASIC(obj)          RBIMPL_CAST((struct RBasic *)(obj))
                                                 ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:25:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/char.h:29:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:30:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:285:1:
error: '__declspec' attributes are not enabled; use '-fdeclspec' or
'-fms-extensions' to enable support for __declspec attributes
RBIMPL_ATTR_NOALIAS()
^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/attr/noalias.h:53:32:
note: expanded from macro 'RBIMPL_ATTR_NOALIAS'
# define RBIMPL_ATTR_NOALIAS() __declspec(noalias)
                               ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:25:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/char.h:29:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:30:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:288:30:
warning: declaration of 'struct RBasic' will not be visible outside of this
function [-Wvisibility]
rbimpl_fl_set_raw_raw(struct RBasic *obj, VALUE flags)
                             ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:290:8:
error: incomplete definition of type 'struct RBasic'
    obj->flags |= flags;
    ~~~^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:288:30:
note: forward declaration of 'struct RBasic'
rbimpl_fl_set_raw_raw(struct RBasic *obj, VALUE flags)
                             ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:298:27:
warning: incompatible pointer types passing 'struct RBasic *' to parameter of
type 'struct RBasic *' [-Wincompatible-pointer-types]
    rbimpl_fl_set_raw_raw(RBASIC(obj), flags);
                          ^~~~~~~~~~~
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:34:30:
note: expanded from macro 'RBASIC'
#define RBASIC(obj)          RBIMPL_CAST((struct RBasic *)(obj))
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/cast.h:33:28:
note: expanded from macro 'RBIMPL_CAST'
# define RBIMPL_CAST(expr) (expr)
                           ^~~~~~
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:288:38:
note: passing argument to parameter 'obj' here
rbimpl_fl_set_raw_raw(struct RBasic *obj, VALUE flags)
                                     ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:310:1:
error: '__declspec' attributes are not enabled; use '-fdeclspec' or
'-fms-extensions' to enable support for __declspec attributes
RBIMPL_ATTR_NOALIAS()
^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/attr/noalias.h:53:32:
note: expanded from macro 'RBIMPL_ATTR_NOALIAS'
# define RBIMPL_ATTR_NOALIAS() __declspec(noalias)
                               ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:25:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/char.h:29:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:30:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:313:32:
warning: declaration of 'struct RBasic' will not be visible outside of this
function [-Wvisibility]
rbimpl_fl_unset_raw_raw(struct RBasic *obj, VALUE flags)
                               ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:315:8:
error: incomplete definition of type 'struct RBasic'
    obj->flags &= ~flags;
    ~~~^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:313:32:
note: forward declaration of 'struct RBasic'
rbimpl_fl_unset_raw_raw(struct RBasic *obj, VALUE flags)
                               ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:323:29:
warning: incompatible pointer types passing 'struct RBasic *' to parameter of
type 'struct RBasic *' [-Wincompatible-pointer-types]
    rbimpl_fl_unset_raw_raw(RBASIC(obj), flags);
                            ^~~~~~~~~~~
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:34:30:
note: expanded from macro 'RBASIC'
#define RBASIC(obj)          RBIMPL_CAST((struct RBasic *)(obj))
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/cast.h:33:28:
note: expanded from macro 'RBIMPL_CAST'
# define RBIMPL_CAST(expr) (expr)
                           ^~~~~~
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:313:40:
note: passing argument to parameter 'obj' here
rbimpl_fl_unset_raw_raw(struct RBasic *obj, VALUE flags)
                                       ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:335:1:
error: '__declspec' attributes are not enabled; use '-fdeclspec' or
'-fms-extensions' to enable support for __declspec attributes
RBIMPL_ATTR_NOALIAS()
^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/attr/noalias.h:53:32:
note: expanded from macro 'RBIMPL_ATTR_NOALIAS'
# define RBIMPL_ATTR_NOALIAS() __declspec(noalias)
                               ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:25:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/char.h:29:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:30:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:338:34:
warning: declaration of 'struct RBasic' will not be visible outside of this
function [-Wvisibility]
rbimpl_fl_reverse_raw_raw(struct RBasic *obj, VALUE flags)
                                 ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:340:8:
error: incomplete definition of type 'struct RBasic'
    obj->flags ^= flags;
    ~~~^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:338:34:
note: forward declaration of 'struct RBasic'
rbimpl_fl_reverse_raw_raw(struct RBasic *obj, VALUE flags)
                                 ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:348:31:
warning: incompatible pointer types passing 'struct RBasic *' to parameter of
type 'struct RBasic *' [-Wincompatible-pointer-types]
    rbimpl_fl_reverse_raw_raw(RBASIC(obj), flags);
                              ^~~~~~~~~~~
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:34:30:
note: expanded from macro 'RBASIC'
#define RBASIC(obj)          RBIMPL_CAST((struct RBasic *)(obj))
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/cast.h:33:28:
note: expanded from macro 'RBIMPL_CAST'
# define RBIMPL_CAST(expr) (expr)
                           ^~~~~~
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:338:42:
note: passing argument to parameter 'obj' here
rbimpl_fl_reverse_raw_raw(struct RBasic *obj, VALUE flags)
                                         ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:465:43:
error: incomplete definition of type 'struct RBasic'
        if (RBASIC_CLASS(x) && !(RBASIC(x)->flags & RUBY_FL_SINGLETON)) {
                                 ~~~~~~~~~^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/fl_type.h:465:34:
note: forward declaration of 'struct RBasic'
        if (RBASIC_CLASS(x) && !(RBASIC(x)->flags & RUBY_FL_SINGLETON)) {
                                 ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rbasic.h:34:50:
note: expanded from macro 'RBASIC'
#define RBASIC(obj)          RBIMPL_CAST((struct RBasic *)(obj))
                                                 ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:25:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic.h:23:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/arithmetic/char.h:29:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:74:19:
error: field has incomplete type 'struct RBasic'
    struct RBasic basic;
                  ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:74:12:
note: forward declaration of 'struct RBasic'
    struct RBasic basic;
           ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:109:26:
error: incomplete definition of type 'struct RBasic'
    VALUE f = RBASIC(str)->flags;
              ~~~~~~~~~~~^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:74:12:
note: forward declaration of 'struct RBasic'
    struct RBasic basic;
           ^
In file included from breakpoint.c:1:
In file included from ./byebug.h:4:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby.h:38:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/ruby.h:26:
In file included from
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core.h:23:
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rarray.h:88:19:
error: field has incomplete type 'struct RBasic'
    struct RBasic basic;
                  ^
/Users/jamiegaskins/.rvm/rubies/ruby-3.0.2/include/ruby-3.0.0/ruby/internal/core/rstring.h:74:12:
note: forward declaration of 'struct RBasic'
    struct RBasic basic;
           ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
6 warnings and 20 errors generated.
make: *** [breakpoint.o] Error 1

make failed, exit code 2

Gem files will remain installed in
/Users/jamiegaskins/.rvm/gems/ruby-3.0.2/gems/byebug-11.1.3 for inspection.
Results logged to
/Users/jamiegaskins/.rvm/gems/ruby-3.0.2/extensions/arm64-darwin-20/3.0.0/byebug-11.1.3/gem_make.out

An error occurred while installing byebug (11.1.3), and Bundler cannot
continue.
Make sure that `gem install byebug -v '11.1.3' --source 'https://rubygems.org/'`
succeeds before bundling.
```

</details>

Maybe `byebug` is meant for Ruby 2.x?

![Byebug: Debugging in Ruby 2](https://user-images.githubusercontent.com/108205/129263226-b4e0ead1-e44d-49d6-81ca-2048f6f37795.png)

But also there hasn't been a release of `byebug` since 8 months before Ruby 3 was released:

![Byebug was last released in April 2020](https://user-images.githubusercontent.com/108205/129263368-cf8f7d65-d141-485b-b241-cfea1c9ac301.png)

![Ruby 3 was released in December 2020](https://user-images.githubusercontent.com/108205/129263475-b6723611-29cf-4d12-8234-506e7623c3c6.png)

citizen428 commented:
@jgaskins There seem to be several people in the issues who are running Byebug under Ruby 3 with various levels of success. Maybe we can try to bundle it off `main` to test this out?

citizen428 commented:
I think in terms of debugging our best bet is probably ditching Byebug and using the new debugger that will be part of Ruby 3.1 (it can already be installed as a gem now):

https://dev.to/st0012/a-sneak-peek-of-ruby-s-new-debugger-5caa

citizen428 commented:
@maestromac I did what I suggested above, replacing `pry-byebug` with the preview of Ruby 3.1's new native debugger.

rhymes commented on Gemfile:
```diff
@@ -141,7 +141,6 @@ group :development, :test do
   gem "bullet", "~> 6.1" # help to kill N+1 queries and unused eager loading
   gem "capybara", "~> 3.35.3" # Capybara is an integration testing tool for rack based web applications
   gem "cypress-rails", "~> 0.5.2" # For end to end tests (E2E)
-  gem "debug", "~> 0.2" # Debugging functionality for Ruby
```
I wonder if this didn't work because it's not pointing to the 1.0 branch line: https://rubygems.org/gems/debug/versions/1.0.0.beta8

citizen428 replied:
Yes, probably. I'm not sure how I came up with that version 🤔 I'll try the 1.0 beta release now.

citizen428 replied:
This seems to have done the trick, thanks 🎉

citizen428 commented:
@rhymes @maestromac We may have to remove `debug` for now still. When trying to start  a Rails console on this branch I get the following error:

```
DEBUGGER: Session start (pid: 22794)
[["DEBUGGER Exception: /home/michi/.asdf/installs/ruby/3.0.2/lib/ruby/gems/3.0.0/gems/debug-1.0.0.beta8/lib/debug/thread_client.rb:643""DEBUGGER Exception: /home/michi/.asdf/installs/ruby/3.0.2/lib/ruby/gems/3.0.0/gems/debug-1.0.0.beta8/lib/debug/thread_client.rb:643",,
```

rhymes commented:
If I start the app with `debug` in it, I get:

```
#<RuntimeError: DEBUGGER: stop at forked process is not supported yet.>#<RuntimeError: DEBUGGER: stop at forked process is not supported yet.>
```

and the app never starts. 

Without it, seems that everything is working correctly! I tested a bit around: posting, uploading images, following people, leaving comments and interacting with the reading list for example

citizen428 commented:
Thanks for confirming, @rhymes. For the time being I removed the `debug` gem again, we should get the debugger automatically once Ruby 3.1 gets released.

citizen428 commented:
Also resolved the merge conflicts related to `docs/.ruby-version` by merging back `main` one more time. Looks like this will finally be ready to go, thanks to everyone who  helped out here, it's been quite the journey.

rhymes reviewed (approved):
LGTM (when SRE and Systems are ready! :))

maestromac reviewed (approved):
Tested. LGTM!

citizen428 commented:
@maestromac I'm not going to make a habit out of this but this is not a PR I'm comfortable merging by myself when nobody else is online and I don't have full systems access in case something goes wrong.
